File: api/internal/handler/dashboard_test.go
package handler

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/RintaroNasu/life-sim/api/internal/httpx"
	"github.com/RintaroNasu/life-sim/api/internal/service"
	"github.com/labstack/echo"
	"github.com/stretchr/testify/require"
)

type fakeDashboardService struct {
	getFunc func(ctx context.Context, userID uint) (*service.DashboardResponse, error)
}

func (f *fakeDashboardService) GetDashboard(ctx context.Context, userID uint) (*service.DashboardResponse, error) {
	if f.getFunc == nil {
		return nil, nil
	}

	return f.getFunc(ctx, userID)
}

func TestDashboardHandler_Get(t *testing.T) {
	tests := []struct {
		name         string
		setUserID    bool
		service      fakeDashboardService
		wantStatus   int
		wantResponse *service.DashboardResponse
		wantAppError *httpx.AppError
	}{
		{
			name:      "【正常系】get成功時は200でダッシュボードデータを返すこと",
			setUserID: true,
			service: fakeDashboardService{
				getFunc: func(ctx context.Context, userID uint) (*service.DashboardResponse, error) {
					require.Equal(t, uint(1), userID)

					return &service.DashboardResponse{
						Year:          2026,
						Month:         8,
						Income:        290000,
						TotalExpenses: 230000,
						FreeAmount:    60000,
						ExpenseBreakdown: []service.ExpenseBreakdownItem{
							{Label: "家賃", Value: 90000},
							{Label: "食費", Value: 50000},
							{Label: "貯金額", Value: 30000},
						},
					}, nil
				},
			},
			wantStatus: http.StatusOK,
			wantResponse: &service.DashboardResponse{
				Year:          2026,
				Month:         8,
				Income:        290000,
				TotalExpenses: 230000,
				FreeAmount:    60000,
				ExpenseBreakdown: []service.ExpenseBreakdownItem{
					{Label: "家賃", Value: 90000},
					{Label: "食費", Value: 50000},
					{Label: "貯金額", Value: 30000},
				},
			},
		},
		{
			name:      "【異常系】contextにuserIDがない場合は401を返すこと",
			setUserID: false,
			wantAppError: httpx.Unauthorized(
				"authenticated user is required",
				errors.New("authenticated user is missing from context"),
			),
		},
		{
			name:      "【異常系】ErrDashboardNotFound時は404を返すこと",
			setUserID: true,
			service: fakeDashboardService{
				getFunc: func(ctx context.Context, userID uint) (*service.DashboardResponse, error) {
					return nil, service.ErrDashboardNotFound
				},
			},
			wantAppError: httpx.NotFound("dashboard data was not found", service.ErrDashboardNotFound),
		},
		{
			name:      "【異常系】想定外エラー時は500を返すこと",
			setUserID: true,
			service: fakeDashboardService{
				getFunc: func(ctx context.Context, userID uint) (*service.DashboardResponse, error) {
					return nil, errors.New("unexpected error")
				},
			},
			wantAppError: httpx.Internal("internal server error", errors.New("unexpected error")),
		},
	}

	for _, tt := range tests {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			e := echo.New()
			req := httptest.NewRequest(http.MethodGet, "/dashboard", nil)
			rec := httptest.NewRecorder()
			c := e.NewContext(req, rec)
			if tt.setUserID {
				c.Set(authUserIDContextKey, uint(1))
			}

			h := NewDashboardHandler(&tt.service)
			err := h.Get(c)

			if tt.wantAppError != nil {
				require.Error(t, err)
				appErr, ok := err.(*httpx.AppError)
				require.True(t, ok)
				require.Equal(t, tt.wantAppError.Status, appErr.Status)
				require.Equal(t, tt.wantAppError.Code, appErr.Code)
				require.Equal(t, tt.wantAppError.Message, appErr.Message)
				return
			}

			require.NoError(t, err)
			require.Equal(t, tt.wantStatus, rec.Code)

			var got service.DashboardResponse
			require.NoError(t, json.Unmarshal(rec.Body.Bytes(), &got))
			require.Equal(t, tt.wantResponse, &got)
		})
	}
}

File: web/lib/api/dashboard.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";
import { getDashboard } from "./dashboard";

afterEach(() => {
  vi.restoreAllMocks();
  vi.unstubAllGlobals();
});

describe("getDashboard", () => {
  it("【正常系】dashboardデータを返すこと", async () => {
    const token = "test-token";
    const responseData = {
      year: 2026,
      month: 8,
      income: 300000,
      total_expenses: 180000,
      free_amount: 120000,
      expense_breakdown: [
        { label: "家賃", value: 90000 },
        { label: "食費", value: 50000 },
      ],
    };

    const fetchMock = vi.fn().mockResolvedValue({
      ok: true,
      json: async () => responseData,
    });

    vi.stubGlobal("fetch", fetchMock);

    const result = await getDashboard(token);

    expect(fetchMock).toHaveBeenCalledWith(
      "http://localhost:8080/dashboard",
      {
        method: "GET",
        headers: {
          Authorization: `Bearer ${token}`,
        },
      },
    );
    expect(result).toEqual(responseData);
  });

  it("【異常系】レスポンスにmessageがある場合はそのmessageでerrorをthrowすること", async () => {
    vi.stubGlobal(
      "fetch",
      vi.fn().mockResolvedValue({
        ok: false,
        status: 404,
        json: async () => ({
          error: {
            message: "dashboard data was not found",
          },
        }),
      }),
    );

    await expect(getDashboard("test-token")).rejects.toMatchObject({
      message: "dashboard data was not found",
      status: 404,
    });
  });

  it("【異常系】レスポンスにmessageがない場合はデフォルトmessageでerrorをthrowすること", async () => {
    vi.stubGlobal(
      "fetch",
      vi.fn().mockResolvedValue({
        ok: false,
        status: 500,
        json: async () => ({
          error: {},
        }),
      }),
    );

    await expect(getDashboard("test-token")).rejects.toMatchObject({
      message: "ダッシュボードデータの取得に失敗しました。",
      status: 500,
    });
  });
});
